pkg/generic/thrift/json.go

```diff
@@ -115,8 +151,75 @@ func (m *ReadJSON) SetBinaryWithBase64(enable bool) {
 	m.binaryWithBase64 = enable
 }
 
+// SetDynamicGo ...
+func (m *ReadJSON) SetDynamicGo(t2jBinaryConv, t2jBinaryConvWithException *t2j.BinaryConv, msg remote.Message) {
+	m.msg = msg
+	m.dynamicgoEnabled = true
+	if m.isClient {
+		// set binary conv to handle an exception field
+		m.t2jBinaryConv = *t2jBinaryConvWithException
+	} else {
+		m.t2jBinaryConv = *t2jBinaryConv
+	}
+}
+
 // Read read data from in thrift.TProtocol and convert to json string
 func (m *ReadJSON) Read(ctx context.Context, method string, in thrift.TProtocol) (interface{}, error) {
+	// fallback logic
+	if !m.dynamicgoEnabled {
+		return m.originalRead(ctx, method, in)
+	}
+
+	// dynamicgo logic
+	tProt, ok := in.(*cthrift.BinaryProtocol)
+	if !ok {
+		return nil, perrors.NewProtocolErrorWithMsg("TProtocol should be BinaryProtocol")
+	}
+
+	var tyDsc *dthrift.TypeDescriptor
+	if m.msg.MessageType() == remote.Reply {
+		tyDsc = m.svc.DynamicGoDsc.Functions()[method].Response()
+	} else {
+		tyDsc = m.svc.DynamicGoDsc.Functions()[method].Request()
+	}
+
+	var resp interface{}
+	if tyDsc.Struct().Fields()[0].Type().Type() == dthrift.VOID {
+		if _, err := tProt.ByteBuffer().ReadBinary(voidWholeLen); err != nil {
+			return nil, err
+		}
+		resp = descriptor.Void{}
+	} else {
+		msgBeginLen := bthrift.Binary.MessageBeginLength(method, thrift.TMessageType(m.msg.MessageType()), m.msg.RPCInfo().Invocation().SeqID())
+		transBuff, err := tProt.ByteBuffer().ReadBinary(m.msg.PayloadLen() - msgBeginLen - bthrift.Binary.MessageEndLength())
+		if err != nil {
+			return nil, err
+		}
+
+		// json size is usually 2 times larger than equivalent thrift data
+		buf := mcache.Malloc(len(transBuff) * 2)[0:0]
+		defer mcache.Free(buf)
```

Comment:
It's dangerous since it's converted to string via `utils.SliceByteToString` which means no copy performed; and buf is returned to mcache, may be modified by another mcache.Malloc caller.

I guess mcache is not suitable here (and other similar places in this pr), @jayantxie what do you think?

Reply:
+1, just use make([]byte...) is enough here, same proble as above

Reply:
fixed